pkg/proxy/metrics/metrics.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// used in the middleware below to capture http status code.
+type statusCatcher struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (s *statusCatcher) WriteHeader(statusCode int) {
+	s.statusCode = statusCode
+	s.ResponseWriter.WriteHeader(statusCode)
+}
+
+// WithRequestCounter counts the number of requests that have passed through
+// the handler it wraps.
+func WithRequestCounter(delegate http.Handler) http.Handler {
```

Comment:
https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/instrument_server.go#L135

Reply:
> // The ObserverVec must have valid metric and label names and must have zero,
// one, or two non-const non-curried labels. For those, the only allowed label
// names are "code" and "method". The function panics otherwise.

I want to come back later and add a shard label.

Reply:
You could just determine the shard at runtime, curry that into the metric, create the delegating handler at runtime and use it?  I'd hate to see us reimplement something that must have been solved as a library with battle-testing behind it :|

Reply:
Oops, didn't notice the comment here https://github.com/kcp-dev/kcp/pull/2302#discussion_r1013368578 which is the same as mine. Sorry